Index: src/stream.c
===================================================================
--- src/stream.c	(revision 15420)
+++ src/stream.c	(working copy)
@@ -295,7 +295,7 @@
         }
       }
     }
-    ices_cue_update (source);
+    //ices_cue_update (source);
   }
 
 #ifdef HAVE_LIBLAME
Index: src/signals.c
===================================================================
--- src/signals.c	(revision 15420)
+++ src/signals.c	(working copy)
@@ -85,7 +85,8 @@
 {
   int stat;
 
-  wait (&stat);
+  while (waitpid (-1, &stat, 0) > 0)
+      ;
 }
 
 /* SIGINT, ok, let's be nice and just drop dead */
Index: src/mp3.c
===================================================================
--- src/mp3.c	(revision 15420)
+++ src/mp3.c	(working copy)
@@ -17,7 +17,7 @@
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
  *
- * $Id: mp3.c,v 1.35 2003/07/06 18:26:47 brendan Exp $
+ * $Id$
  */
 
 #include "definitions.h"
@@ -26,7 +26,7 @@
 
 #define MPG_MD_MONO 3
 
-#define MP3_BUFFER_SIZE 4096
+#define MP3_BUFFER_SIZE 1500
 
 typedef struct {
   unsigned int version;
@@ -256,7 +256,7 @@
 {
   ices_mp3_in_t* mp3_data = self->data;
   int remaining;
-  int rlen = 0;
+  int rlen = -1;
 
   if (mp3_data->buf) {
     remaining = mp3_data->len - mp3_data->pos;
@@ -278,7 +278,8 @@
       rlen = read (self->fd, buf, len);
   }
 
-  self->bytes_read += rlen;
+  if (rlen > 0)
+      self->bytes_read += rlen;
 
   return rlen;
 }
