diff -u src/mp3.c src.new/mp3.c
--- src/mp3.c	2009-08-08 22:04:06.000000000 +0100
+++ src.new/mp3.c	2014-01-01 13:48:40.178160888 +0000
@@ -26,7 +26,7 @@
 
 #define MPG_MD_MONO 3
 
-#define MP3_BUFFER_SIZE 4096
+#define MP3_BUFFER_SIZE 1500
 
 typedef struct {
 	unsigned int version;
@@ -272,9 +272,10 @@
 			rlen = read(self->fd, buf, len);
 	}
 
-	self->bytes_read += rlen;
+        if (rlen > 0)
+            self->bytes_read += rlen;
 
-	return rlen;
+	return rlen == 0 ? -1 : rlen;
 }
 
 #ifdef HAVE_LIBLAME
diff -u src/signals.c src.new/signals.c
--- src/signals.c	2009-08-06 21:00:27.000000000 +0100
+++ src.new/signals.c	2014-01-01 13:48:00.601483613 +0000
@@ -81,7 +81,8 @@
 static RETSIGTYPE signals_child(const int sig) {
 	int stat;
 
-	wait(&stat);
+	while (waitpid (-1, &stat, 0) > 0)
+           ;
 }
 
 /* SIGINT, ok, let's be nice and just drop dead */
diff -u src/stream.c src.new/stream.c
--- src/stream.c	2009-08-08 21:56:38.000000000 +0100
+++ src.new/stream.c	2014-01-01 13:49:17.098859698 +0000
@@ -294,7 +294,7 @@
 				}
 			}
 		}
-		ices_cue_update(source);
+		// ices_cue_update(source);
 		if ( source->interrupttime && time(NULL)>=source->interrupttime ) finish_send = 1;
 	}
 
