diff -u src/mp3.c src.new/mp3.c
--- src/mp3.c	2009-08-08 22:04:06.000000000 +0100
+++ src.new/mp3.c	2013-09-20 22:29:22.772755438 +0100
@@ -267,12 +267,13 @@
 	} else {
 		/* we don't just use EOF because we'd like to avoid the ID3 tag */
 		if (self->filesize && self->filesize - self->bytes_read < len)
-			len = self->filesize - self->bytes_read;
+			rlen = self->filesize - self->bytes_read;
 		if (len)
 			rlen = read(self->fd, buf, len);
 	}
 
-	self->bytes_read += rlen;
+        if (rlen > 0)
+            self->bytes_read += rlen;
 
 	return rlen;
 }
Common subdirectories: src/playlist and src.new/playlist
diff -u src/signals.c src.new/signals.c
--- src/signals.c	2009-08-06 21:00:27.000000000 +0100
+++ src.new/signals.c	2013-09-20 22:30:23.181341124 +0100
@@ -81,7 +81,7 @@
 static RETSIGTYPE signals_child(const int sig) {
 	int stat;
 
-	wait(&stat);
+	while (waitpid (-1, &stat, 0) > 0);
 }
 
 /* SIGINT, ok, let's be nice and just drop dead */
diff -u src/stream.c src.new/stream.c
--- src/stream.c	2009-08-08 21:56:38.000000000 +0100
+++ src.new/stream.c	2013-09-20 22:29:44.935603457 +0100
@@ -294,7 +294,7 @@
 				}
 			}
 		}
-		ices_cue_update(source);
+		//ices_cue_update(source);
 		if ( source->interrupttime && time(NULL)>=source->interrupttime ) finish_send = 1;
 	}
 
