Index: src/id3.c
===================================================================
--- src/id3.c	(revision 19189)
+++ src/id3.c	(working copy)
@@ -237,7 +237,8 @@
 
   /* ices_log_debug("ID3v2: Frame type [%s] found, %d bytes", hdr, len); */
   if (!strcmp (hdr, ID3V2_ARTIST_TAG(tag)) || !strcmp (hdr, ID3V2_TITLE_TAG(tag))) {
-    if (! (buf = malloc(len+1))) {
+    char *b;
+    if (! (b = buf = malloc(len+1))) {
       ices_log ("Error allocating memory while reading ID3v2 frame");
       
       return -1;
@@ -244,14 +245,15 @@
     }
     len2 = len;
     while (len2) {
-      if ((rlen = source->read (source, buf, len)) < 0) {
+      if ((rlen = source->read (source, b, len2)) <= 0) {
         ices_log ("Error reading ID3v2 frame data");
         free (buf);
-      
+
         return -1;
       }
       tag->pos += rlen;
       len2 -= rlen;
+      b += rlen;
     }
 
     /* skip encoding */
Index: src/mp3.c
===================================================================
--- src/mp3.c	(revision 19189)
+++ src/mp3.c	(working copy)
@@ -17,7 +17,7 @@
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
  *
- * $Id: mp3.c,v 1.35 2003/07/06 18:26:47 brendan Exp $
+ * $Id$
  */
 
 #include "definitions.h"
@@ -26,7 +26,7 @@
 
 #define MPG_MD_MONO 3
 
-#define MP3_BUFFER_SIZE 4096
+#define MP3_BUFFER_SIZE 1500
 
 typedef struct {
   unsigned int version;
@@ -278,7 +278,8 @@
       rlen = read (self->fd, buf, len);
   }
 
-  self->bytes_read += rlen;
+  if (rlen > 0)
+      self->bytes_read += rlen;
 
   return rlen;
 }
Index: src/signals.c
===================================================================
--- src/signals.c	(revision 19189)
+++ src/signals.c	(working copy)
@@ -85,7 +85,8 @@
 {
   int stat;
 
-  wait (&stat);
+  while (waitpid (-1, &stat, 0) > 0)
+      ;
 }
 
 /* SIGINT, ok, let's be nice and just drop dead */
Index: src/stream.c
===================================================================
--- src/stream.c	(revision 19189)
+++ src/stream.c	(working copy)
@@ -295,7 +295,7 @@
         }
       }
     }
-    ices_cue_update (source);
+    //ices_cue_update (source);
   }
 
 #ifdef HAVE_LIBLAME
